kaspresso/src/main/kotlin/com/kaspersky/kaspresso/device/bluetooth/BluetoothImpl.kt

```diff
@@ -0,0 +1,120 @@
+package com.kaspersky.kaspresso.device.bluetooth
+
+import android.bluetooth.BluetoothManager
+import android.content.Context
+import com.kaspersky.components.kautomator.system.UiSystem
+import com.kaspersky.kaspresso.device.server.AdbServer
+import com.kaspersky.kaspresso.flakysafety.algorithm.FlakySafetyAlgorithm
+import com.kaspersky.kaspresso.internal.exceptions.AdbServerException
+import com.kaspersky.kaspresso.internal.systemscreen.NotificationsFullScreen
+
+import com.kaspersky.kaspresso.logger.UiTestLogger
+import com.kaspersky.kaspresso.params.FlakySafetyParams
+
+/**
+ * The implementation of the [Bluetooth] interface.
+ */
+class BluetoothImpl(
+    private val logger: UiTestLogger,
+    private val targetContext: Context,
+    private val adbServer: AdbServer
+) : Bluetooth {
+
+    companion object {
+        private const val CMD_STATE_ENABLE = "enable"
+        private const val CMD_STATE_DISABLE = "disable"
+        private const val BLUETOOTH_STATE_CHANGE_CMD = "svc bluetooth"
+        private const val BLUETOOTH_STATE_CHANGE_ROOT_CMD = "su 0 svc bluetooth"
+        private const val BLUETOOTH_STATE_CHECK_CMD = "settings get global bluetooth_on"
+        private const val BLUETOOTH_STATE_CHECK_RESULT_ENABLED = "1"
+        private const val BLUETOOTH_STATE_CHECK_RESULT_DISABLED = "0"
+        private val ADB_RESULT_REGEX = Regex("exitCode=(\\d+), message=(.+)")
+    }
+
+    private val flakySafetyAlgorithm = FlakySafetyAlgorithm(logger)
+    private val flakySafetyParams: FlakySafetyParams
+        get() = FlakySafetyParams(
+            timeoutMs = 1000,
+            intervalMs = 100,
+            allowedExceptions = setOf(AdbServerException::class.java)
+        )
+
+    override fun enable() {
+        toggleBluetooth(enable = true)
+    }
+
+    override fun disable() {
+        toggleBluetooth(enable = false)
+    }
+
+    /**
+     * Toggles Bluetooth state
+     * Tries, first and foremost, to send ADB command. If this attempt fails,
+     * opens Android Settings screen and tries to switch Bluetooth setting thumb.
+     */
+    private fun toggleBluetooth(enable: Boolean) {
+        if (isBluetoothNotSupported()) {
+            logger.i("Bluetooth is not supported")
+            return
+        }
+        logger.i("${if (enable) "En" else "Dis"}able bluetooth")
```

Comment:
Please, make two separate calls to logger. Concatenating log strings make them almost imposible to find in code